cpp/src/arrow/ipc/writer.cc

```diff
@@ -449,14 +453,22 @@ class RecordBatchSerializer {
 
   template <typename T>
   enable_if_base_binary<typename T::TypeClass, Status> Visit(const T& array) {
+    using offset_type = typename T::offset_type;
+
     std::shared_ptr<Buffer> value_offsets;
     RETURN_NOT_OK(GetZeroBasedValueOffsets<T>(array, &value_offsets));
     auto data = array.value_data();
 
     int64_t total_data_bytes = 0;
     if (value_offsets) {
-      total_data_bytes = array.value_offset(array.length()) - array.value_offset(0);
+      offset_type last_offset_value;
+      RETURN_NOT_OK(MemoryManager::CopyBufferSliceToCPU(
+          value_offsets, array.length() * sizeof(offset_type), sizeof(offset_type),
+          reinterpret_cast<uint8_t*>(&last_offset_value)));
+
+      total_data_bytes = last_offset_value;
```

Comment:
Why don't we take into account offset #0 here anymore?

Reply:
just above this we use `GetZeroBasedValueOffsets` and populate `value_offsets` with zero-indexed offsets. This means that if we grab the very last offset in that buffer (`array.length() * sizeof(offset_type)`) we have the full number of data bytes and we can save the extra copy to grab offset #0